### DISCUSSION TITLE
How to send KernelArguments towards KernelFunctions when using IChatCompletionService

### DISCUSSION THREAD

**Discussion by ystvan** [microsoft/semantic-kernel#6457] [Q&A] [answered]
My goal is that the user can have a conversation with the model through Semantic Kernel's AzureOpenAIChatCompletion service.
I register two plugins as Kernel Functions that might be triggered during the conversation. I need to set the `securityFilters` value on runtime.
 
I cannot seem to find a way to initialize a Kernel Arguments arguments like

```
 var kernelArguments = new KernelArguments()
 {
     ["securityFilters"] = "some-string"
 };
```

An that could be used in a plugin as 

```
[KernelFunction, Description("Provide document look up in company data for employees using SharePoint")]
[return: Description("document's filename, source as hyperlink and its content")]
public async Task<List<IndexSchema>> SearchAsync(
    [Description("The user intent in a a LUCENE search query format")] string query,
    [Description("The SharePoint security group memberships of the user")] string securityFilters)
{    
        return await _searchService.SearchAsync("index-01", query, securityFilter);
}
```
Where do I pass the KernelArguments into?

If it was a one-time prompt I could easily use `InvokePromptStreamingAsync` method that accepts KernelArguments, but I need the ChatCompletionService to maintain chat history, etc.

This is a rudimentary example where I struggle to pass in the KernelArguments :)
```
 var kernelBuilder = Kernel.CreateBuilder();

 kernelBuilder.AddAzureOpenAIChatCompletion(
          "model",
          "endpoint",
          "key");

 kernelBuilder.Plugins.AddFromType<AzureAISearchPlugin>();

 // Create kernel
 var kernel = kernelBuilder.Build();

 ChatHistory history = [];
 
 // Get chat completion service
 _chatCompletionService = kernel.GetRequiredService<IChatCompletionService>();

 // Enable auto function calling
 OpenAIPromptExecutionSettings openAIPromptExecutionSettings = new()
 {
     MaxTokens = 4000,
     ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions
 };

 await foreach (var chatUpdate in _chatCompletionService.GetStreamingChatMessageContentsAsync(
                history,
                executionSettings: openAIPromptExecutionSettings,
                kernel: kernel))
 {
     yield return new TokenizedResponse(chatUpdate.Content);

     responseBuffer.Append(chatUpdate.Content);
 }
```

**Comment by dmytrostruk** [comment, marked as answer]
@ystvan Thank you for your question!

There are a couple of ways how to achieve your scenario:
1. If `securityFilters` property does not contain PII or any other sensitive information, in your chat history you can inject these details as part of user or system message. Since auto function invocation is enabled, LLM will be able to use that information and Semantic Kernel will inject it into your plugin.
2. If you do not wish to share `securityFilters` content with LLM, you could use `kernel.Data` property bag to set this information. In your plugin, you can inject kernel instance, which should contain that property, here is an example of `Kernel` injection:
https://github.com/microsoft/semantic-kernel/blob/8d9c3cbc996a5feea6485b2c21169e9d7558ad68/dotnet/samples/Concepts/Functions/MethodFunctions_Types.cs#L178-L184

Please let us know if any of these approaches work for you, thanks!

**Reply by ystvan** [reply]
Amazing! Did not know about the `kernel.Data` property, it came to the rescue! Thank you @dmytrostruk
